[compiler-rt] Replace direct calls to pipe with internal_pipe

Lancern opened the pull request:
This PR tries to resolve google/sanitizers#1106 by introducing a new `internal_pipe` function which will not be intercepted by TSAN.

cjappl commented:
This seems reasonable to me. 

Please:
* Fix the code formatting issue
* Add a test that catches this issue and proves it doesn't happen anymore

I'm going to add one of the original commenters on your issue to this PR.

Lancern committed:
[compiler-rt] Replace direct calls to pipe with internal_pipe

This patch tries to resolve google/sanitizers#1106 by introducing a new
internal_pipe function which will not be intercepted by TSAN.

Lancern commented:
@cjappl Sorry for the delay here. Updated this PR and added a test that verify TSAN does not complain any more.
